src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,427 @@
+import { ResourceStore } from './ResourceStore';
+import streamifyArray from 'streamify-array';
+import { UnsupportedMediaTypeHttpError } from '../util/errors/UnsupportedMediaTypeHttpError';
+import { CONTENT_TYPE_QUADS, DATA_TYPE_BINARY, DATA_TYPE_QUAD } from '../util/ContentTypes';
+import { createReadStream, createWriteStream, promises as fsPromises, Stats } from 'fs';
+import { DataFactory, StreamParser, StreamWriter } from 'n3';
+import { ensureTrailingSlash, trimTrailingSlashes } from '../util/Util';
+import { extname, join as joinPath } from 'path';
+import { LDP, RDF, STAT, TERMS, XML } from '../util/Prefixes';
+import { NamedNode, Quad } from 'rdf-js';
+
+/**
+ * Resource store storing its data in the file system backend.
+ * All requests will throw an {@link NotFoundHttpError} if unknown identifiers get passed.
+ */
+export class FileResourceStore implements ResourceStore {
+  private readonly baseRequestURI: string;
+  private readonly rootFilepath: string;
+  private readonly interactionController: InteractionController;
+
+  private readonly predicates = {
+    aType: DataFactory.namedNode(`${RDF}type`),
+    modified: DataFactory.namedNode(`${TERMS}modified`),
+    contains: DataFactory.namedNode(`${LDP}contains`),
+    mtime: DataFactory.namedNode(`${STAT}mtime`),
+    size: DataFactory.namedNode(`${STAT}size`),
+  };
+
+  private readonly objects = {
+    container: DataFactory.namedNode(`${LDP}Container`),
+    basicContainer: DataFactory.namedNode(`${LDP}BasicContainer`),
+    ldpResource: DataFactory.namedNode(`${LDP}Resource`),
+    dateTime: DataFactory.namedNode(`${XML}dateTime`),
+  };
+
+  /**
+   * @param baseRequestURI - Will be stripped of all incoming URIs and added to all outgoing ones to find the relative
+   * path.
+   * @param rootFilepath - Root filepath in which the resources and containers will be saved as files and directories.
+   * @param interactionController - Instance of InteractionController to use.
+   */
+  public constructor(baseRequestURI: string, rootFilepath: string, interactionController: InteractionController) {
+    this.baseRequestURI = trimTrailingSlashes(baseRequestURI);
+    this.rootFilepath = trimTrailingSlashes(rootFilepath);
+    this.interactionController = interactionController;
+  }
+
+  /**
+   * Store the incoming data as a file under a file path corresponding to `container.path`,
+   * where slashes correspond to subdirectories.
+   * @param container - The identifier to store the new data under.
+   * @param representation - Data to store. Only File streams are supported.
+   *
+   * @returns The newly generated identifier.
+   */
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    if (representation.dataType !== DATA_TYPE_BINARY) {
+      throw new UnsupportedMediaTypeHttpError('FileResourceStore only supports binary representations.');
+    }
+    const path = this.parseIdentifier(container);
```

Comment:
Probably need a comment above this line, describing in one line what this block does. (If not one line, then maybe we need two blocks.)